### PR TITLE
Laboratory work #4, Andrey Gusev - 22FPL1 

### PR DESCRIPTION
**Pull request by AndreyGusev1**

### CODE REVIEW

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +30,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +30,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+        text_words = text.replace('!', f' {self._end_of_word_token} ').replace('?', f' {self._end_of_word_token} ')\
+            .replace('.', f' {self._end_of_word_token} ').lower().split()
+        tokens = []
+        for word in text_words:
+            if word.isalpha() or word.isspace() or word == self._end_of_word_token:
+                tokens.append(word)
+            else:
+                letters = []
+                for symbol in word:
+                    if symbol.isalpha():
```
[the remark this answers is not in the document]
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -39,6 +57,11 @@ def _put(self, element: str) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(element, str) or not element:
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -98,6 +134,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if not isinstance(seq_len, int) or not isinstance(prompt, str) or seq_len < 0 or not prompt:
+            raise ValueError
+        text = self._word_processor.encode(prompt)
+        if not text:
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -98,6 +134,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        count = 0
+        while count < seq_len:
+            word = self._model.generate_next_token(text)
+            if word is None:
+                raise ValueError
```
[the remark this answers is not in the document]
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: seminars/practice_10_legb.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# pylint: disable=missing-function-docstring
```
Comment (artyomtugaryov) [inline seminars/practice_10_legb.py:1]: revert!

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -28,6 +30,20 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError('Text cannot be empty')
+        text_words = text.replace('!', f' {self._end_of_word_token} ').replace('?', f' {self._end_of_word_token} ')\
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:35]: too difficult, split

File: lab_4_fill_words_by_ngrams/main.py [hunk 8]
```diff
@@ -56,6 +77,16 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            raise ValueError
+        decoded_corpus = ' '.join(decoded_corpus).replace(f' {self._end_of_word_token}', '.')
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:82]: `f' {self._end_of_word_token}'` to a variable, please

File: lab_4_fill_words_by_ngrams/main.py [hunk 9]
```diff
@@ -98,6 +132,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            word = self._model.generate_next_token(text)
+            if word is None:
+                raise ValueError('The word is None')
+            if not word:
+                break
+            word = sorted(word.items(), key=lambda x: (x[1], x[0]), reverse=True)
+            sum_ = 0
+            tokens = []
+            for token in word:
+                if sum_ < self._p_value:
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:151]: else - break?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py [hunk 10]
```diff
@@ -98,6 +132,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+                    sum_ += token[1]
+                    tokens.append(token[0])
+            randomized = random.choice(tokens)
+            text += (randomized, )
+            count += 1
+        decoded = self._word_processor.decode(text)
+        if decoded is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:159]: add  message

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
The same comments

File: lab_4_fill_words_by_ngrams/main.py (hunk 7 above)
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 8 above)
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 9 above)
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py:153]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 10 above)
Comment (AndreyGusev1) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**